### PR TITLE
Migrating dependency management to go modules

### PR DESCRIPTION
**Pull request by mattfarina**
This PR migrates from dep to go modules for dependency management.

Some notes:

* Helm is now outside the GOPATH for the working directory. This causes go modules to automatically be in use for all go commands
* The cache has been switched to go modules. When the GOPATH is deprecated where will these things be cached?
* `go mod` manages the `go.mod` file and sorts things in ascii alpha order. If you change that and use a `go mod edit` command it will re-sort

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

File: go.mod
Comment (bacongobbler) [inline go.mod]: What do you think about moving all the direct dependencies of Helm to the top of the `require` file, `goimports` style (separated by an empty newline)? That way we can clearly separate direct and indirect dependencies. We'd still want to keep them in alphabetical order, of course. :)

Kinda like this:

```
require (
  github.com/BurntSushi/toml v0.3.1
  github.com/Masterminds/semver v1.4.2
  github.com/Masterminds/vcs v1.13.0
  ...

  cloud.google.com/go v0.34.0 // indirect
  contrib.go.opencensus.io/exporter/ocagent v0.2.0 // indirect
)
```

**Review by bacongobbler — APPROVED** [review]
changes LGTM, just had one comment about the go.mod file

I'd hold off on merging until @adamreese has a look at this as well. This is a major change and will affect everyone's development workflow once merged.

File: go.mod
Comment (mattfarina) [reply go.mod]: If you notice, the dependencies are listed in alpha (ascii) sorted order. If I were to run a command like:

```
$ go mod -require github.com/Masterminds/semver@v1.4.2
```

The command would rewrite the file and put it back in alpha order. To control the ordering means we would need to manually manage the file all the time.

I like the idea of ordering it our way but I don't know that I want to tell people they can't use `go mod edit` to manage it.

**Comment by mattfarina** [comment]
For reference, the original PR tool about 11 minutes to popular the cache. With a populated cache the test section of CircleCI took about 40 seconds.

File: go.mod
Comment (bacongobbler) [reply go.mod]: bleh. :( okay. I'm not too familiar with `go mod` yet. Thanks for the heads up!

File: docs/developers.md
Comment (hickeyma) [inline docs/developers.md]: Should this note be updated because I think by using Go modules `GOPATH` is no longer mandatory? If so then update in install doc as well.

**Review by hickeyma — COMMENTED** [review]
This is looking good. Have a small question about doc update.

Also, I tried to do a `make build` locally in path outside of `GOPATH`. I am getting the following error:
```console
$ make build
go build  -tags '' -ldflags '-w -s -X helm.sh/helm/internal/version.gitCommit=d4323c1da8dd1c18f0460fbece45765e259dba9d -X helm.sh/helm/internal/version.gitTreeState=dirty' -o /home/usr1/test/helm/bin/helm helm.sh/helm/cmd/helm
go: finding github.com/garyburd/redigo v1.6.0
[..........]
go: finding k8s.io/client-go/util latest
go: finding k8s.io/client-go v11.0.0+incompatible
go: downloading k8s.io/client-go v11.0.0+incompatible
go: extracting github.com/gophercloud/gophercloud v0.0.0-20190126172459-c818fa66e4c8
go: extracting k8s.io/client-go v11.0.0+incompatible
go: extracting sigs.k8s.io/kustomize v2.0.3+incompatible
build helm.sh/helm/cmd/helm: cannot load k8s.io/api/admissionregistration/v1alpha1: cannot find module providing package k8s.io/api/admissionregistration/v1alpha1
Makefile:56: recipe for target '/home/usr1/test/helm/bin/helm' failed
make: *** [/home/usr1/test/helm/bin/helm] Error 1
```
Just wonder if there is an issue/clash with existing go modules?

File: Makefile
Comment (ian-howell) [inline Makefile]: While we're making the switch to go modules, we should begin tracking dependencies. This guarantees a reproducible environment for future builds, as well as relieving the strain of manual dependency management for developers.

Documentation on doing so is here: https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module

File: Makefile
Comment (unguiculus) [inline Makefile]: In module-aware mode, this will add `gox` as a dependency to `go.mod` which is probably not desired. Could be fixed like this:

`GO111MODULE=off go get -u github.com/mitchellh/gox`

Comment (ian-howell) [reply Makefile]: On the contrary, I think adding `gox` as a dependency to `go.mod` is ideal, as it guarantees that every developer is using the same version of the tool. See [here](https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module) for a more detailed explanation on how this would work.

Comment (unguiculus) [reply Makefile]: Interesting. Thanks for the link. I wasn't aware of that. Then, I guess, a `tools.go` file should be added as suggested.

File: go.mod
Comment (unguiculus) [inline go.mod]: I'm still trying to wrap my head around some of the module concepts, especially versioning rules.

Wouldn't this have to be `module helm.sh/helm/v3`?

See:
https://github.com/golang/go/wiki/Modules#semantic-import-versioning
https://github.com/golang/go/wiki/Modules#releasing-modules-v2-or-higher

Comment (ankushchadha) [reply go.mod]: yes, the suffix /v3 needs to be added to the module name. And all the import paths need to refer to the new module name (with suffix /v3). If these two steps are avoided, then this module will be referred to as helm.sh/helm v3.0.0+incompatible

Note - if you are using non-module aware static analyzer tools, then they will fail because such tools will expect v3 folder on the filesystem under GOPATH.

**Comment by bacongobbler** [comment]
So, I think we should consider moving this forward to unblock @hickeyma in #6154. Do you have time to work on this before the beta release, or would you like someone else to pick this up? :)

**Comment by ankushchadha** [comment]
Let us know if we can help with this PR. This is exactly what we do right now i.e. help open source projects adopt Go Modules. To begin with, the change includes adding /v3 suffix to the module name and update all the import path (main and test code). Apart from this change, if code generators and static code analysis tools are used, then they need to be updated as well.

**Comment by hickeyma** [comment]
@bacongobbler Re https://github.com/helm/helm/pull/5498#issuecomment-517752821, no longer blocked. Had issues more when PoCing and using local versions of the repo. All is ok now.

**Comment by markmssd** [comment]
I've got some experience with GO modules, @mattfarina can I help on anything there? Other than the Merge Conflicts, was there anything else to do?

**Comment by sbose78** [comment]
This would be really good to have!

**Comment by sagikazarmark** [comment]
We used to have such a cache in place in our builds, but we removed it and replaced it with GOPROXY for a number of reasons:

- the cache is append only, meaning it grows over time. We started with 1.2 GB and ended up with 2 GB after a few months.
- The download/upload duration was not consistent. Sometimes it took 4 seconds, sometimes a minute (or more).

**Comment by Ilyes512** [comment]
This really should be added, and I hope that happens before v3 goes stable.

**Comment by markmssd** [comment]
Totally agree. As a workaround I use it through GO's `exec.Command`, pretty hacky but it works

**Comment by gkarthiks** [comment]
totally @markmssd, but I would like to interact directly from my program to helm modules.

**Comment by mattfarina** [comment]
I'll resolve the conflicts and update this sometime next week.

**Comment by cagataygurturk** [comment]
Looking forward to this change in Helm 3

**Comment by mattfarina** [comment]
The updates are now ready for review

File: Makefile
Comment (bacongobbler) [inline Makefile]: how come we need to `cd` to the root here? Is this a thing required for CI?

File: go.mod
Comment (bacongobbler) [reply go.mod]: does this still need addressing?

File: Makefile
Comment (mattfarina) [reply Makefile]: if you run `go get` from within the project directory, even to install an app like gox, it will add it to the `go.mod` file and deal with the dependency tree for it. So, run `build-cross` locally and have it install gox and you've modified your `go.mod` file with gox and an updated dependency tree.

Comment (bacongobbler) [reply Makefile]: Ahh. I see what you mean. Thanks!

**Comment by mattfarina** [comment]
Yes... the module paths need to be fixed. So many places need to be updated to have a `/v3` added. Just spotted that.

**Review by bacongobbler — APPROVED** [review]

**Review by thomastaylor312 — APPROVED** [review]

**Comment by karuppiah7890** [comment]
Finally! Yay!!

**Comment by cagataygurturk** [comment]
@mattfarina Thanks for the PR! Just trying to include dev-v3 branch into my project but I get the following error. Any help?

```
$ go get -t  helm.sh/helm@dev-v3
go: finding helm.sh/helm dev-v3
go get helm.sh/helm@dev-v3: helm.sh/helm@v0.0.0-20191003184432-6f0235f6595c: invalid version: go.mod has post-v0 module path "helm.sh/helm/v3" at revision 6f0235f6595c
```

**Comment by sagikazarmark** [comment]
@cagataygurturk see https://github.com/helm/helm/pull/5498#issuecomment-538060650

**Comment by mattfarina** [comment]
@cagataygurturk That's a good question. Go modules are different from inclusion in the past once the major version goes beyond 1. I did a quick writeup at https://codeengineered.com/blog/2019/go-mod-major-versions/.

You need to run to import v3 from `helm.sh/helm/v3`. But, this isn't going to work right now because we need to fix the redirects on helm.sh. I filed https://github.com/helm/helm-www/issues/300 to record the issue. If no one gets to it first I'll try to find time next week to do this.

**Pull request merged**
